pkg/koordlet/runtimehooks/nri/server.go

```diff
@@ -0,0 +1,190 @@
+	DisableStages map[string]struct{}
+	Executor      resourceexecutor.ResourceUpdateExecutor
+}
+
+type NriServer struct {
+	stub            stub.Stub
+	mask            stub.EventMask
+	options         Options // server options
+	runPodSandbox   func(*NriServer, *api.PodSandbox, *api.Container) error
+	createContainer func(*NriServer, *api.PodSandbox, *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error)
+	updateContainer func(*NriServer, *api.PodSandbox, *api.Container) ([]*api.ContainerUpdate, error)
+}
+
+var (
+	_ = stub.ConfigureInterface(&NriServer{})
+	_ = stub.SynchronizeInterface(&NriServer{})
+	_ = stub.RunPodInterface(&NriServer{})
+	_ = stub.CreateContainerInterface(&NriServer{})
+	_ = stub.UpdateContainerInterface(&NriServer{})
+
+	pluginName = "koordlet_nri"
+	pluginIdx  = "00"
+	events     = "RunPodSandbox,CreateContainer,UpdateContainer"
+	cfg        nriconfig
+	opts       []stub.Option
+	err        error
+)
+
+func NewNriServer(opt Options) (*NriServer, error) {
+	opts = append(opts, stub.WithPluginName(pluginName))
+	opts = append(opts, stub.WithPluginIdx(pluginIdx))
+	opts = append(opts, stub.WithSocketPath(filepath.Join(system.Conf.VarRunRootDir, "nri/nri.sock")))
+	p := &NriServer{options: opt}
+	if p.mask, err = api.ParseEventMask(events); err != nil {
+		klog.Errorf("failed to parse events: %v", err)
+	}
+	cfg.Events = strings.Split(events, ",")
+
+	if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+		klog.Errorf("failed to create plugin stub: %v", err)
+	}
+
+	return p, err
+}
+
+func (s *NriServer) Start() error {
+	go func() {
+		if s.stub != nil {
+			err := s.stub.Run(context.Background())
+			if err != nil {
+				klog.Errorf("nri server exited with error %v", err)
+			} else {
+				klog.Info("nri server start")
+			}
+		} else {
+			klog.Error("nri stub is nil")
+		}
+	}()
+	return nil
+}
+
+func (p *NriServer) Configure(config, runtime, version string) (stub.EventMask, error) {
+	klog.Infof("got configuration data: %q from runtime %s %s", config, runtime, version)
+	if config == "" {
+		return p.mask, nil
+	}
+
+	err := yaml.Unmarshal([]byte(config), &cfg)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse provided configuration: %w", err)
+	}
+
+	p.mask, err = api.ParseEventMask(cfg.Events...)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse events in configuration: %w", err)
+	}
+
+	return p.mask, nil
+}
+
+func (p *NriServer) Synchronize(pods []*api.PodSandbox, containers []*api.Container) ([]*api.ContainerUpdate, error) {
+	// todo: update existed containers configure
+	return nil, nil
+}
+
+func (p *NriServer) RunPodSandbox(pod *api.PodSandbox) error {
+	podCtx := &protocol.PodContext{}
+	podCtx.FromNri(pod)
+	// todo: return error or bypass error based on PluginFailurePolicy
+	err := hooks.RunHooks(p.options.PluginFailurePolicy, rmconfig.PreRunPodSandbox, podCtx)
+	if err != nil {
+		klog.Errorf("hooks run error: %v", err)
```

Comment:
We'd better handle the error according to the failure policy. e.g. throw the error to the nri runtime when the `policy=Fail`

Reply:
ok, for these fatal errors, we can reture them to containerd